cpp/extract_normdata_translations.cc

```diff
@@ -0,0 +1,225 @@
+        record.extractSubfields(tag, "9" , &ixtheo_lang_code);
+        for (auto lang_code : ixtheo_lang_code) {
+            if (lang_code.find("eng") != std::string::npos && *ixtheo_pos != "IxTheo_eng")
+               *ixtheo_pos = *ixtheo_pos + "_eng";
+            // FIXME: There are currently no french translations... 
+            else if (lang_code.find("fra") != std::string::npos && *ixtheo_pos != "IxTheo_fra")
+               *ixtheo_pos = *ixtheo_pos + "_fra";
+        }
+    }
+}
+
+
+void ExtractTranslations(File* const marc_norm_input, 
+                         const std::string german_term_field_spec,
+                         const std::string translation_field_spec,
+                         std::map<std::string,std::string> term_to_translation_map[]) {
+
+    std::set<std::string> german_tags_and_subfield_codes;
+    if (unlikely(StringUtil::Split(german_term_field_spec, ':', &german_tags_and_subfield_codes) < 1))
+        Error("ExtractTranslations: Need at least one translation field");
+
+    std::set<std::string> translation_tags_and_subfield_codes;
+    if (unlikely(StringUtil::Split(translation_field_spec, ':', &translation_tags_and_subfield_codes) < 1))
+        Error("ExtractTranslations: Need at least one translation field");
+
+    unsigned count(0);
+    
+    while (const MarcUtil::Record record = MarcUtil::Record::XmlFactory(marc_norm_input)) {
+
+        std::vector<std::string> german_term;
+
+        // Determine the german term we will have translations for
+
+        for (auto tag_and_subfields : german_tags_and_subfield_codes) {
+             const std::string tag(tag_and_subfields.substr(0, 3));
+             const std::string subfields(tag_and_subfields.substr(3));
+             
+             std::vector<std::string> german_term_for_one_field;
+             record.extractSubfields(tag, subfields, &german_term_for_one_field);
+             
+             // We may get the german term from only one field
+             if (german_term.size() > 1)  
+                 Warning("We have german terms in more than one field for PPN: " + record.getFields()[0]);
+
+             if (not german_term_for_one_field.empty()) 
+                 german_term = german_term_for_one_field;
+        }        
+
+        std::vector<std::string> all_translations;
+
+        // Extract all additional translations
+
+        for (auto tag_and_subfields : translation_tags_and_subfield_codes) {
+            const std::string tag(tag_and_subfields.substr(0, 3));
+            const std::string subfields(tag_and_subfields.substr(3));
+
+            std::vector<std::string> translations;
+            record.extractSubfields(tag, subfields, &translations);
+
+            // For IxTheo-Translations add the language code in the same field
+            augment_ixtheo_tag_with_language(record, tag, translations);
+            
+            all_translations.insert(std::end(all_translations), std::begin(translations), std::end(translations));
```

Comment:
Instead of "std::end()" please use the cbegin() and cend() member functions of std::vector.  "std::end" is only needed when you want to write templates that apart from certain container classes also work with built in arrays.